Sources/GRPC/GRPCKeepaliveHandlers.swift

```diff
@@ -0,0 +1,354 @@
+  let pingHandler: PingHandler
+
+  /// The scheduled task which will ping.
+  var scheduledPing: Scheduled<Void>? = nil
+
+  /// The scheduled task which will close the connection.
+  var scheduledClose: Scheduled<Void>? = nil
+
+  func close(context: ChannelHandlerContext) {
+    context.close(mode: .all, promise: nil)
+  }
+}
+
+protocol _ChannelKeepaliveHandler: ChannelInboundHandler {
+  var pingHandler: PingHandler { get }
+  var scheduledPing: Scheduled<Void>? { get set }
+  var scheduledClose: Scheduled<Void>? { get set }
+  func close(context: ChannelHandlerContext)
+}
+
+extension _ChannelKeepaliveHandler where OutboundOut == HTTP2Frame, InboundIn == HTTP2Frame {
+  func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+    if event is NIOHTTP2StreamCreatedEvent {
+      self.perform(action: self.pingHandler.streamCreated(), context: context)
+    } else if event is StreamClosedEvent {
+      self.perform(action: self.pingHandler.streamClosed(), context: context)
+    }
+
+    context.fireUserInboundEventTriggered(event)
+  }
+
+  func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    switch self.unwrapInboundIn(data).payload {
+    case let .ping(pingData, ack: ack):
+      self.perform(action: self.pingHandler.read(pingData: pingData, ack: ack), context: context)
+    default:
+      break
+    }
+
+    context.fireChannelRead(data)
+  }
+
+  private func perform(action: PingHandler.Action, context: ChannelHandlerContext) {
+    switch action {
+    case let .schedulePing(delay, timeout):
+      self.schedulePing(delay: delay, timeout: timeout, context: context)
+    case .cancelScheduledPing:
+      self.cancelScheduledPing()
+    case .cancelScheduledTimeout:
+      self.cancelScheduledTimeout()
+    case let .reply(payload):
+      self.send(payload: payload, context: context)
+    case .none:
+      break
+    }
+  }
+
+  private func send(payload: HTTP2Frame.FramePayload, context: ChannelHandlerContext) {
+    let frame = self.wrapOutboundOut(.init(streamID: .rootStream, payload: payload))
+    context.writeAndFlush(frame, promise: nil)
+  }
+
+  private func schedulePing(delay: TimeAmount, timeout: TimeAmount, context: ChannelHandlerContext) {
+    guard delay != .nanoseconds(Int64.max) else { return }
+
+    self.scheduledPing = context.eventLoop.scheduleTask(in: delay) { [unowned self] in
+      self.perform(action: self.pingHandler.pingFired(), context: context)
+      self.scheduleClose(timeout: timeout, context: context)
+    }
+  }
+
+  private func scheduleClose(timeout: TimeAmount, context: ChannelHandlerContext) {
+    self.scheduledClose = context.eventLoop.scheduleTask(in: timeout) { [unowned self] in
+      self.close(context: context)
+    }
+  }
+
+  private func cancelScheduledPing() {
+    self.scheduledPing?.cancel()
+    self.scheduledPing = nil
+  }
+
+  private func cancelScheduledTimeout() {
+    self.scheduledClose?.cancel()
+    self.scheduledClose = nil
+  }
+}
```

Comment:
Can we add an implementation of `handlerRemoved` that also cancels the two timeouts? That'll help with ensuring that any strong reference cycles really are broken.